abci/types/types.go

```diff
@@ -167,3 +170,125 @@ func RespondVerifyVoteExtension(ok bool) ResponseVerifyVoteExtension {
+	s := make([][]byte, len(r))
+	for i, e := range r {
+		d := deterministicExecTxResult(e)
+		b, err := d.Marshal()
+		if err != nil {
+			panic(err)
+		}
+		s[i] = b
+	}
+	return s
+}
+
+func (tr *TxRecord) IsIncluded() bool {
+	return tr.Action == TxRecord_ADDED || tr.Action == TxRecord_UNMODIFIED
+}
+
+// IncludedTxs returns all of the TxRecords that are marked for inclusion in the
+// proposed block.
+func (rpp *ResponsePrepareProposal) IncludedTxs() []*TxRecord {
+	trs := []*TxRecord{}
+	for _, tr := range rpp.TxRecords {
+		if tr.IsIncluded() {
+			trs = append(trs, tr)
+		}
+	}
+	return trs
+}
+
+// RemovedTxs returns all of the TxRecords that are marked for removal from the
+// mempool.
+func (rpp *ResponsePrepareProposal) RemovedTxs() []*TxRecord {
+	trs := []*TxRecord{}
+	for _, tr := range rpp.TxRecords {
+		if tr.Action == TxRecord_REMOVED {
+			trs = append(trs, tr)
+		}
+	}
+	return trs
+}
+
+// AddedTxs returns all of the TxRecords that are marked as added to the proposal.
+func (rpp *ResponsePrepareProposal) AddedTxs() []*TxRecord {
+	trs := []*TxRecord{}
+	for _, tr := range rpp.TxRecords {
+		if tr.Action == TxRecord_ADDED {
+			trs = append(trs, tr)
+		}
+	}
+	return trs
+}
+
+// Validate checks that the fields of the ResponsePrepareProposal are properly
+// constructed. Validate returns an error if any of the validation checks fail.
+func (rpp *ResponsePrepareProposal) Validate(maxSizeBytes int64, otxs [][]byte) error {
+
+	if !rpp.ModifiedTx {
+		// This method currently only checks the validity of the TxRecords field.
+		// If ModifiedTx is false, then we can ignore the validity of the TxRecords field.
+		//
+		// TODO: When implementing VoteExensions, AppSignedUpdates may be modified by the application
+		// and this method should be updated to validate the AppSignedUpdates.
+		return nil
+	}
+
+	// TODO: this feels like a large amount allocated data. We move all the Txs into strings
+	// in the map. The map will be as large as the original byte slice.
+	// Is there a key we can use?
+	otxsSet := make(map[string]struct{}, len(otxs))
+	for _, tx := range otxs {
+		otxsSet[string(tx)] = struct{}{}
+	}
+	ntx := map[string]struct{}{}
+	var size int64
+	for _, tr := range rpp.TxRecords {
+		if tr.IsIncluded() {
+			size += int64(len(tr.Tx))
+			if _, ok := ntx[string(tr.Tx)]; ok {
+				return errors.New("duplicate included transaction")
+			}
+			ntx[string(tr.Tx)] = struct{}{}
+		}
+		if _, ok := otxsSet[string(tr.Tx)]; ok {
+			if tr.Action == TxRecord_ADDED {
+				return fmt.Errorf("unmodified transaction incorrectly marked as %s", tr.Action.String())
+			}
+		} else {
+			if tr.Action == TxRecord_REMOVED || tr.Action == TxRecord_UNMODIFIED {
+				return fmt.Errorf("new transaction incorrectly marked as %s", tr.Action.String())
+			}
+		}
+		if tr.Action == TxRecord_UNKNOWN {
+			return fmt.Errorf("transaction incorrectly marked as %s", tr.Action.String())
+		}
+	}
+	if size > maxSizeBytes {
```

Comment:
I'd recommend doing this check inside the loop. If we exceed it, there's no need to make further copies of transaction data later in the slice since we're only going to fail anyway.